r/src/materialize.c

```diff
@@ -21,258 +21,215 @@
-  const uint8_t* data_buffer = array_view->buffer_views[1].data.as_uint8;
-
-  // Fill the buffer
-  switch (array_view->storage_type) {
-    case NANOARROW_TYPE_BOOL:
-      for (R_xlen_t i = 0; i < array_view->array->length; i++) {
-        result[i] = ArrowBitGet(data_buffer, i);
-      }
-
-      // Set any nulls to NA_LOGICAL
-      if (is_valid != NULL && array_view->array->null_count != 0) {
-        for (R_xlen_t i = 0; i < array_view->array->length; i++) {
-          if (!ArrowBitGet(is_valid, i)) {
-            result[i] = NA_LOGICAL;
-          }
-        }
-      }
+// Needed for the list_of materializer
+#include "convert.h"
+
+#include "materialize.h"
+#include "materialize_blob.h"
+#include "materialize_chr.h"
+#include "materialize_date.h"
+#include "materialize_dbl.h"
+#include "materialize_difftime.h"
+#include "materialize_int.h"
+#include "materialize_lgl.h"
+#include "materialize_posixct.h"
+#include "materialize_unspecified.h"
+
+SEXP nanoarrow_alloc_type(enum VectorType vector_type, R_xlen_t len) {
+  SEXP result;
+
+  switch (vector_type) {
+    case VECTOR_TYPE_LGL:
+      result = PROTECT(Rf_allocVector(LGLSXP, len));
       break;
-    case NANOARROW_TYPE_INT8:
-    case NANOARROW_TYPE_UINT8:
-    case NANOARROW_TYPE_INT16:
-    case NANOARROW_TYPE_UINT16:
-    case NANOARROW_TYPE_INT32:
-    case NANOARROW_TYPE_UINT32:
-    case NANOARROW_TYPE_INT64:
-    case NANOARROW_TYPE_UINT64:
-    case NANOARROW_TYPE_FLOAT:
-    case NANOARROW_TYPE_DOUBLE:
-      for (R_xlen_t i = 0; i < array_view->array->length; i++) {
-        result[i] = ArrowArrayViewGetIntUnsafe(array_view, i) != 0;
-      }
-
-      // Set any nulls to NA_LOGICAL
-      if (is_valid != NULL && array_view->array->null_count != 0) {
-        for (R_xlen_t i = 0; i < array_view->array->length; i++) {
-          if (!ArrowBitGet(is_valid, i)) {
-            result[i] = NA_LOGICAL;
-          }
-        }
-      }
+    case VECTOR_TYPE_INT:
+      result = PROTECT(Rf_allocVector(INTSXP, len));
+      break;
+    case VECTOR_TYPE_DBL:
+      result = PROTECT(Rf_allocVector(REALSXP, len));
+      break;
+    case VECTOR_TYPE_CHR:
+      result = PROTECT(Rf_allocVector(STRSXP, len));
       break;
-
     default:
-      UNPROTECT(1);
       return R_NilValue;
   }
 
   UNPROTECT(1);
-  return result_sexp;
+  return result;
 }
 
-SEXP nanoarrow_materialize_int(struct ArrowArrayView* array_view) {
-  SEXP result_sexp = PROTECT(Rf_allocVector(INTSXP, array_view->array->length));
-  int* result = INTEGER(result_sexp);
-  int64_t n_bad_values = 0;
-
-  // True for all the types supported here
-  const uint8_t* is_valid = array_view->buffer_views[0].data.as_uint8;
+int nanoarrow_ptype_is_data_frame(SEXP ptype) {
+  return Rf_isObject(ptype) && TYPEOF(ptype) == VECSXP &&
+         (Rf_inherits(ptype, "data.frame") ||
+          (Rf_xlength(ptype) > 0 && Rf_getAttrib(ptype, R_NamesSymbol) != R_NilValue));
```

Comment:
This `getAttrib()` might materialize the row names vector, which can be avoided.